crates/ra_hir/src/code_model.rs

```diff
@@ -1078,11 +1078,32 @@ impl Type {
         )
     }
 
-    // FIXME: this method is broken, as it doesn't take closures into account.
-    pub fn as_callable(&self) -> Option<CallableDef> {
-        Some(self.ty.value.as_callable()?.0)
+    pub fn as_function(&self) -> Option<Function> {
+        match self.ty.value.as_callable()? {
+            CallableDef::FunctionId(id) => Some(id.into()),
+            _ => None
+        }
+    }
+
+    pub fn as_tuple_or_unit_struct(&self) -> Option<Struct> {
+        match self.ty.value.as_callable()? {
+            CallableDef::StructId(id) => Some(id.into()),
+            _ => None
+        }
     }
 
+    pub fn as_tuple_or_unit_enum_variant(&self) -> Option<EnumVariant> {
```

Comment:
Hm, I think you can't actually invoke a unit struct or enum... I wonder if we should add a check for unit structs here:
https://github.com/rust-analyzer/rust-analyzer/blob/93527b5f193ed214f6ae0f8112eaec2eebabd016/crates/ra_ide/src/display/function_signature.rs#L59